cumulusci/tasks/bulkdata/load.py

```diff
@@ -138,41 +139,43 @@ def _execute_step(
 
         mapping["oid_as_pk"] = bool(mapping.get("fields", {}).get("Id"))
 
+        local_ids = []
+        query = self._query_db(mapping)
+        operation = (
+            DataOperationType.INSERT
+            if mapping.get("action") == "insert"
+            else DataOperationType.UPDATE
+        )
```

Comment:
Perhaps the mapping parser should be updated to parse `action` into one of these two types.

Reply:
Yeah, this one is also held back by the legacy code that constructs mapping steps as dicts (to support `after` steps). This ugliness can be refactored out when I fix that.